### PR TITLE
 Fix: A VM would not restart if the init process crashes

### LINKED ISSUE #196: Unmanaged vm state when a vm exits unexpectedly
**Describe the bug**
When a vm exits undexpectedly the supervisord stuck on "connection refused" all the time
the vm is always considered as ready to receive requests

**To Reproduce**

## deploy main.py
```
import os
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}

@app.get("/fail")
async def fail():
    os._exit(0)
    return {"message": "failed"}
```
curl vm_host/.../ => OK
curl vm_host/../fail => Connection closed
curl vm_host/.../ => Failed

### PR DESCRIPTION
**Pull request by hoh**
Fixes  #196

Do not squash

### CODE REVIEW

File: vm_supervisor/run.py [hunk 1]
```diff
@@ -44,6 +44,37 @@ async def build_event_scope(event) -> Dict[str, Any]:
...
+        raise HTTPBadRequest(reason="Code, runtime or data not available")
+    except FileTooLargeError as error:
+        raise HTTPInternalServerError(reason=error.args[0])
+    except VmSetupError as error:
+        logger.exception(error)
+        pool.forget_vm(vm_hash=vm_hash)
+        raise HTTPInternalServerError(reason="Error during program initialisation")
+    except MicroVMFailedInit as error:
+        logger.exception(error)
+        pool.forget_vm(vm_hash=vm_hash)
```
Comment (odesenfans) [inline vm_supervisor/run.py:69]: These two lines are duplicated in most cases. Not a big deal.

File: vm_supervisor/run.py [hunk 2]
```diff
@@ -94,6 +94,18 @@ async def run_code_on_request(
     try:
         await execution.becomes_ready()
         result_raw: bytes = await execution.run_code(scope=scope)
+
+        if result_raw == b'':
+            # Missing result from the init process of the virtual machine, not even an error message.
+            # It may have completely crashed.
+
+            # Stop the virtual machine with due to failing init.
```
Comment (odesenfans) [inline vm_supervisor/run.py:102]: "with due"?

File: vm_supervisor/run.py [hunk 3]
```diff
@@ -94,6 +94,18 @@ async def run_code_on_request(
...
+            # It will be restarted on a future request.
+            await execution.stop()
+
+            return web.Response(status=502, reason="No response from VM",
+                                text="VM did not respond and was shutdown")
```
Comment (odesenfans) [inline vm_supervisor/run.py:107]: shutdown -> shut down.

**Review by odesenfans — APPROVED** [review]
Minor comments and questions.

**Review by aliel — APPROVED** [review]

File: vm_supervisor/run.py (hunk 2 above)
Comment (hoh) [reply vm_supervisor/run.py]: Fixed and squashed

File: vm_supervisor/run.py (hunk 3 above)
Comment (hoh) [reply vm_supervisor/run.py]: Fixed and squashed

File: vm_supervisor/run.py [hunk 4]
```diff
@@ -44,6 +44,37 @@ async def build_event_scope(event) -> Dict[str, Any]:
...
+        execution = await pool.create_a_vm(
+            vm_hash=vm_hash,
+            program=message.content,
+            original=original_message.content,
+        )
+    except ResourceDownloadError as error:
+        logger.exception(error)
+        pool.forget_vm(vm_hash=vm_hash)
+        raise HTTPBadRequest(reason="Code, runtime or data not available")
+    except FileTooLargeError as error:
```
[the remark this answers is not in the document]
Comment (hoh) [reply vm_supervisor/run.py:61]: Fixed in 228805c90ea4f83dccfa94828942238b686c896e

**Pull request merged**